src/Control/Monad/Bayes/Sampler.hs

```diff
@@ -43,97 +44,127 @@ import Control.Monad.Bayes.Class
         uniform
       ),
   )
-import Control.Monad.ST (ST, runST, stToIO)
-import Control.Monad.State (State, state)
-import Control.Monad.Trans (MonadIO, lift)
-import Control.Monad.Trans.Reader (ReaderT, ask, mapReaderT, runReaderT)
+import Control.Monad.ST (ST)
+import Control.Monad.Trans (lift)
+import Control.Monad.Trans.Reader (ReaderT, ask, runReaderT)
 import Data.Fixed (mod')
-import Numeric.Log (Log (ln))
-import System.Random.MWC
-  ( Gen,
-    GenIO,
-    GenST,
-    Seed,
-    Variate (uniform, uniformR),
-    create,
-    createSystemRandom,
-    restore,
-    save,
-  )
+import Numeric.Log (Log (..))
 import System.Random.MWC.Distributions qualified as MWC
+import System.Random.Stateful (IOGenM, STGenM, StatefulGen, StdGen, mkStdGen, newIOGenM, newSTGenM, uniformRM)
 
 -- | An 'IO' based random sampler using the MWC-Random package.
-newtype SamplerIO a = SamplerIO (ReaderT GenIO IO a)
-  deriving newtype (Functor, Applicative, Monad, MonadIO)
+newtype SamplerIO g a = SamplerIO (StatefulGen g IO => ReaderT g IO a)
```

Comment:
Why not just keep this compositional, as `Sampler g IO a` or (changing the variable order) `Sampler IO g a`?

Reply:
That is, why not avoid the type synonym `SamplerIO` altogether?

Reply:
I think it was for backward compatibility but I am happy to remove it.